core/src/main/java/org/elasticsearch/action/support/master/TransportMasterNodeAction.java

```diff
@@ -81,146 +83,150 @@ protected void processBeforeDelegationToMaster(Request request, ClusterState sta
+                        @Override
+                        public Response newInstance() {
+                            return newResponse();
+                        }
+
+                        @Override
+                        public void handleResponse(Response response) {
+                            listener.onResponse(response);
+                        }
+
+                        @Override
+                        public String executor() {
+                            return ThreadPool.Names.SAME;
+                        }
+
+                        @Override
+                        public void handleException(final TransportException exp) {
+                            Throwable cause = exp.unwrapCause();
+                            if (cause instanceof ConnectTransportException) {
+                                // we want to retry here a bit to see if a new master is elected
+                                logger.debug("connection exception while trying to forward request to master node [{}], scheduling a retry. Error: [{}]",
+                                        nodes.masterNode(), exp.getDetailedMessage());
+                                retry(cause, masterNodeChangedPredicate());
+                            } else {
+                                listener.onFailure(exp);
                             }
-                    );
+                        }
+                    });
                 }
-                return;
             }
-            processBeforeDelegationToMaster(request, clusterState);
-            transportService.sendRequest(nodes.masterNode(), actionName, request, new BaseTransportResponseHandler<Response>() {
+        }
+
+        private ClusterStateObserver.ChangePredicate retryableOrNoBlockPredicate() {
+            return new ClusterStateObserver.ValidationPredicate() {
                 @Override
-                public Response newInstance() {
-                    return newResponse();
+                protected boolean validate(ClusterState newState) {
+                    ClusterBlockException blockException = checkBlock(request, newState);
+                    return (blockException == null || !blockException.retryable());
                 }
+            };
+        }
 
+        private ClusterStateObserver.ChangePredicate masterNodeChangedPredicate() {
+            return new ClusterStateObserver.ChangePredicate() {
                 @Override
-                public void handleResponse(Response response) {
-                    listener.onResponse(response);
+                public boolean apply(ClusterState previousState, ClusterState.ClusterStateStatus previousStatus,
+                                     ClusterState newState, ClusterState.ClusterStateStatus newStatus) {
+                    return newState.nodes().masterNodeId() != null && (previousState != newState || previousStatus != newStatus);
```

Comment:
I know this is how it used to be, but can we make the if be more like the `masterNodeChangePredicate` name and check the the master node is not null and have changed? (we now test for a cluster state change)